include/openPMD/IO/ADIOS/ParallelADIOS1IOHandler.hpp

```diff
@@ -22,39 +22,80 @@
+#   include <adios.h>
+#   include <adios_read.h>
 #endif
 
 #include <future>
 #include <memory>
 #include <string>
+#if openPMD_HAVE_ADIOS1
+#   include <unordered_map>
+#   include <unordered_set>
+#endif
+
+#if _MSC_VER
+#   define EXPORT __declspec( dllexport )
+#else
+#   define EXPORT __attribute__((visibility("default")))
+#endif
 
 
 namespace openPMD
 {
 #if openPMD_HAVE_ADIOS1 && openPMD_HAVE_MPI
-class ParallelADIOS1IOHandler;
+class EXPORT ParallelADIOS1IOHandler;
 
-class ParallelADIOS1IOHandlerImpl : public ADIOS1IOHandlerImpl
+class EXPORT ParallelADIOS1IOHandlerImpl : public AbstractIOHandlerImpl
 {
 public:
     ParallelADIOS1IOHandlerImpl(AbstractIOHandler*, MPI_Comm);
     virtual ~ParallelADIOS1IOHandlerImpl();
 
-    virtual void init() override;
+    virtual void init();
 
     virtual std::future< void > flush() override;
 
-    virtual int64_t open_write(Writable *) override;
-    virtual ADIOS_FILE* open_read(std::string const& name) override;
-
-private:
+    virtual void createFile(Writable*, Parameter< Operation::CREATE_FILE > const&) override;
+    virtual void createPath(Writable*, Parameter< Operation::CREATE_PATH > const&) override;
+    virtual void createDataset(Writable*, Parameter< Operation::CREATE_DATASET > const&) override;
+    virtual void extendDataset(Writable*, Parameter< Operation::EXTEND_DATASET > const&) override;
+    virtual void openFile(Writable*, Parameter< Operation::OPEN_FILE > const&) override;
+    virtual void openPath(Writable*, Parameter< Operation::OPEN_PATH > const&) override;
+    virtual void openDataset(Writable*, Parameter< Operation::OPEN_DATASET > &) override;
+    virtual void deleteFile(Writable*, Parameter< Operation::DELETE_FILE > const&) override;
+    virtual void deletePath(Writable*, Parameter< Operation::DELETE_PATH > const&) override;
+    virtual void deleteDataset(Writable*, Parameter< Operation::DELETE_DATASET > const&) override;
+    virtual void deleteAttribute(Writable*, Parameter< Operation::DELETE_ATT > const&) override;
+    virtual void writeDataset(Writable*, Parameter< Operation::WRITE_DATASET > const&) override;
+    virtual void writeAttribute(Writable*, Parameter< Operation::WRITE_ATT > const&) override;
+    virtual void readDataset(Writable*, Parameter< Operation::READ_DATASET > &) override;
+    virtual void readAttribute(Writable*, Parameter< Operation::READ_ATT > &) override;
+    virtual void listPaths(Writable*, Parameter< Operation::LIST_PATHS > &) override;
+    virtual void listDatasets(Writable*, Parameter< Operation::LIST_DATASETS > &) override;
+    virtual void listAttributes(Writable*, Parameter< Operation::LIST_ATTS > &) override;
+
+    virtual int64_t open_write(Writable *);
+    virtual ADIOS_FILE* open_read(std::string const& name);
+    void close(int64_t);
+    void close(ADIOS_FILE*);
+
+protected:
+    int64_t m_group;
+    std::string m_groupName;
+    ADIOS_READ_METHOD m_readMethod;
```

Comment:
If we don't use that varaible for both backends anymore, we cna probably hard-code the ```ADIOS_READ_METHOD``` used in each one.

Reply:
I don't fully get this but I guess we can do this in a follow-up as well.

Reply:
The methods are initialized once per handler in ```init()```
https://github.com/openPMD/openPMD-api/blob/265828b7f79428e2cf3caa3e9252b03c0f29f1c9/src/IO/ADIOS/ADIOS1IOHandler.cpp#L210
https://github.com/openPMD/openPMD-api/blob/551ab6df4c3ad27b5cbee7cbf9db888fd503d37c/src/IO/ADIOS/ParallelADIOS1IOHandler.cpp#L221-L222
With the hard separation of serial and parallel backend, we do not really need the variable any more.